# MLflow results presentation (2)

In [58]:
import pandas as pd
from uncertainties import ufloat
import mlflow
from mlflow.tracking import MlflowClient
mlflow.set_tracking_uri('file:///home/zeyusun/work/flare-prediction-smarp/mlruns')
client = MlflowClient()

from mlflow_helper import *

pd.set_option('display.max_colwidth', None)

%load_ext tensorboard
%load_ext autoreload
%autoreload 2

The tensorboard extension is already loaded. To reload it, use:
  %reload_ext tensorboard
The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [59]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style>

## Compare leaderboard 3: LSTM_CNN vs base

In [ ]:
def compare(base_run_name, target_run_name):
    base_exp_name, target_exp_name = ['leaderboard3'] * 2
    if '/' in base_run_name:
        base_exp_name, base_run_name = base_run_name.split('/')
    if '/' in target_run:
        target_exp_name, target_run_name = target_run_name.split('/')

    runs_raw_1 = retrieve('leaderboard3', 'LSTM_CNN')
    runs_1 = select(runs_raw_1)
    organizes

In [81]:
runs_raw_1 = retrieve('leaderboard3', 'LSTM_CNN')
runs_1 = select(runs_raw_1)

ckpt = runs_raw_1['tags.checkpoint'].str.extract(r'epoch=(?P<epoch>[0-9]+)-step=(?P<step>[0-9]+)').astype(int)
runs_1 = pd.concat((runs_1, ckpt), axis=1)

runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
207 2021-06-27 16:03:33.239000+00:00            LSTM_CNN   

                tags.mlflow.source.git.commit  
207  e13810ffddafdb818df7c4de68d9eafd67861e8c  


In [83]:
organize(runs_1, std=True)

dataset                  fused_sharp          fused_smarp  \
      estimator                                             
ACC   CNN              0.943+/-0.030        0.910+/-0.019   
      LSTM             0.957+/-0.012        0.921+/-0.016   
AUC   CNN              0.983+/-0.013        0.964+/-0.012   
      LSTM             0.988+/-0.004        0.970+/-0.010   
TSS   CNN              0.886+/-0.059        0.820+/-0.038   
      LSTM             0.913+/-0.025        0.841+/-0.032   
HSS   CNN              0.886+/-0.059        0.820+/-0.038   
      LSTM             0.913+/-0.025        0.841+/-0.032   
BSS   CNN              0.343+/-0.388        0.365+/-0.305   
      LSTM             0.793+/-0.055        0.621+/-0.056   
epoch CNN              6.600+/-4.980        9.200+/-7.530   
      LSTM            10.600+/-7.537       16.200+/-2.588   
step  CNN         1301.600+/-878.957  1582.200+/-1135.646   
      LSTM       1933.800+/-1228.537   2716.000+/-436.486   

dataset                        sharp               smarp  
      estimator                                           
ACC   CNN              0.934+/-0.011       0.888+/-0.021  
      LSTM             0.951+/-0.012       0.918+/-0.017  
AUC   CNN              0.981+/-0.004       0.956+/-0.016  
      LSTM             0.986+/-0.005       0.965+/-0.009  
TSS   CNN              0.867+/-0.023       0.776+/-0.041  
      LSTM             0.901+/-0.024       0.836+/-0.034  
HSS   CNN              0.867+/-0.023       0.776+/-0.041  
      LSTM             0.901+/-0.024       0.836+/-0.034  
BSS   CNN              0.619+/-0.125       0.274+/-0.493  
      LSTM             0.757+/-0.062       0.581+/-0.047  
epoch CNN            28.000+/-19.545     16.000+/-10.223  
      LSTM             8.200+/-7.190      17.000+/-1.225  
step  CNN        1668.800+/-1263.607  1567.200+/-962.466  
      LSTM         491.600+/-356.530  1664.400+/-109.703

In [84]:
runs_raw_2 = retrieve('leaderboard3', 'base')
runs_2 = select(runs_raw_2)

ckpt = runs_raw_2['tags.checkpoint'].str.extract(r'epoch=(?P<epoch>[0-9]+)-step=(?P<step>[0-9]+)').astype(int)
runs_2 = pd.concat((runs_2, ckpt), axis=1)

runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
124 2021-08-18 03:00:51.425000+00:00                base   

                tags.mlflow.source.git.commit  
124  98d231a7d80767545d12d384c881f6eb545ffb7a  


In [85]:
organize(runs_2, std=True)

dataset                 fused_sharp         fused_smarp              sharp  \
      estimator                                                              
ACC   CNN             0.892+/-0.076       0.839+/-0.072      0.772+/-0.081   
      LSTM            0.944+/-0.012       0.909+/-0.026      0.934+/-0.023   
AUC   CNN             0.980+/-0.010       0.968+/-0.006      0.967+/-0.029   
      LSTM            0.990+/-0.005       0.971+/-0.010      0.985+/-0.004   
TSS   CNN             0.784+/-0.152       0.678+/-0.143      0.545+/-0.162   
      LSTM            0.888+/-0.023       0.818+/-0.052      0.867+/-0.046   
HSS   CNN             0.784+/-0.152       0.678+/-0.143      0.545+/-0.162   
      LSTM            0.888+/-0.023       0.818+/-0.052      0.867+/-0.046   
BSS   CNN             0.569+/-0.359       0.376+/-0.286      0.101+/-0.346   
      LSTM            0.797+/-0.052       0.642+/-0.103      0.736+/-0.076   
epoch CNN             1.400+/-1.517       2.600+/-2.191     10.200+/-4.712   
      LSTM            5.200+/-3.421       5.600+/-3.362      3.600+/-2.881   
step  CNN         407.000+/-263.260   569.800+/-344.752  608.800+/-196.895   
      LSTM       1041.400+/-557.124  1022.000+/-482.548  250.400+/-153.295   

dataset                      smarp  
      estimator                     
ACC   CNN            0.879+/-0.014  
      LSTM           0.911+/-0.015  
AUC   CNN            0.958+/-0.014  
      LSTM           0.970+/-0.009  
TSS   CNN            0.758+/-0.028  
      LSTM           0.822+/-0.030  
HSS   CNN            0.758+/-0.028  
      LSTM           0.822+/-0.030  
BSS   CNN            0.544+/-0.049  
      LSTM           0.629+/-0.049  
epoch CNN            6.400+/-3.847  
      LSTM           7.400+/-1.817  
step  CNN        689.200+/-368.656  
      LSTM       780.400+/-187.720

#### Sharp problem
* What: In `base` run, CNN trained on SHARP withseed 3 has particularly bad performance.
* Why: AUC is a bad metric. The step 494 (end of epoch 8) has higher validation AUC than its neighboring epochs (not very distinguishable, 0.994 vs 0.986 and 0.989), but other metrics are lower (e.g., 0.702 vs 0.840 and 0.883). Looking at SSP, the curve significantly tilted leftwards, indicating a significant amount of underforecasting.
* This also suggests using an optimized TSS is a dangerous idea because it can select a under- or over-forecasting model that is not robust to the threshold.
    * It is very much possible that the test SSP is a bit different from the validation SSP, and using the threshold fitted on the validation set may lead to a performance much lower than expected. This is due to the sensitivity of SSP to the threshold.
    * Thresholding on the test set may not be very useful in practice, as we are not allowed to give a prediction of a test instance unless we know the entire test set.
* Why: AUC is not consistent with others is because it is the area under the uniform measure of false alarm rate, which is rarely the case in practice where false alarm rate is selected to be close to zero. AUC gives too much weight on the part where false alarm is high.
    * Area Under SSP may be a good idea. It is the area using the uniform measure of threshold.

In [88]:
runs_comb = pd.concat((
    runs_raw_1.loc[[198]],
    runs_raw_2.loc[[115]],
))
runs_comb

run_id experiment_id    status  \
198  f7bc8479849e4a30805eb7cfd116395e            15  FINISHED   
115  dc8a1de8327b49838879eb7ec444feed            15  FINISHED   

                                                                                              artifact_uri  \
198  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/f7bc8479849e4a30805eb7cfd116395e/artifacts   
115  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/dc8a1de8327b49838879eb7ec444feed/artifacts   

                          start_time                         end_time  \
198 2021-06-27 16:49:13.611000+00:00 2021-06-27 17:03:02.065000+00:00   
115 2021-08-18 03:39:51.850000+00:00 2021-08-18 03:48:29.301000+00:00   

     metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
198          0.979329             0.953096                     0.926554   
115          0.916192             0.307692                     0.860640   

     metrics.test/hss1  ...  tags.mlflow.source.name  \
198           0.853659  ...           sweep_arnet.py   
115           0.300188  ...           sweep_arnet.py   

              tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
198  0d9e66f8c3ca4266a4ca374e4d3b377c                    LOCAL           True   
115  b168fac075e243be9bc28f888ef21a23                    LOCAL           True   

     tags.mlflow.user  tags.database_name  \
198           zeyusun            M_Q_24hr   
115           zeyusun            M_Q_24hr   

                tags.mlflow.source.git.commit  tags.estimator_name  \
198  e13810ffddafdb818df7c4de68d9eafd67861e8c                  CNN   
115  98d231a7d80767545d12d384c881f6eb545ffb7a                  CNN   

                                                                                              tags.checkpoint  \
198  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_953/checkpoints/epoch=34-step=1924.ckpt   
115   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1275/checkpoints/epoch=8-step=494.ckpt   

     tags.mlflow.runName  
198   M_Q_24hr_CNN_sharp  
115   M_Q_24hr_CNN_sharp  

[2 rows x 81 columns]

In [89]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

Reusing TensorBoard on port 6008 (pid 32279), started 8:22:49 ago. (Use '!kill 32279' to kill it.)

#### Fused_smarp problem
* What: fused_smarp is worse than smarp in all metrics except for AUC. LSTM_CNN has a similar AUC to base, but other metrics are much better than base.
* Why: We think maybe it's because base has threahold 0.5 is the problem.
    * In `base` run, CNN trained on fused_smarp with seed 2 is a perfect piece of evidence that AUC is a bad validation metric. According to AUC, epoch 2 has is the best. But SSP is severely tilted. Let's use TSS.

In [86]:
runs_comb = pd.concat((
    runs_raw_1.loc[[170]],
    runs_raw_2.loc[[87]],
))
runs_comb

run_id experiment_id    status  \
170  ea083959ee884688966dbaeac8079b08            15  FINISHED   
87   fabff25c06b34516b1b229b6c45ca1b4            15  FINISHED   

                                                                                              artifact_uri  \
170  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/ea083959ee884688966dbaeac8079b08/artifacts   
87   file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/fabff25c06b34516b1b229b6c45ca1b4/artifacts   

                          start_time                         end_time  \
170 2021-06-27 21:00:49.510000+00:00 2021-06-27 21:21:07.944000+00:00   
87  2021-08-18 07:28:38.833000+00:00 2021-08-18 07:42:04.967000+00:00   

     metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
170          0.968733             0.892280                     0.864382   
87           0.967358             0.473968                     0.784749   

     metrics.test/hss1  ...  tags.mlflow.source.name  \
170           0.825853  ...           sweep_arnet.py   
87            0.464093  ...           sweep_arnet.py   

              tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
170  0d9e66f8c3ca4266a4ca374e4d3b377c                    LOCAL           True   
87   b168fac075e243be9bc28f888ef21a23                    LOCAL           True   

     tags.mlflow.user  tags.database_name  \
170           zeyusun            M_Q_24hr   
87            zeyusun            M_Q_24hr   

                tags.mlflow.source.git.commit  tags.estimator_name  \
170  e13810ffddafdb818df7c4de68d9eafd67861e8c                  CNN   
87   98d231a7d80767545d12d384c881f6eb545ffb7a                  CNN   

                                                                                             tags.checkpoint  \
170  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_981/checkpoints/epoch=9-step=1649.ckpt   
87   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1303/checkpoints/epoch=2-step=494.ckpt   

          tags.mlflow.runName  
170  M_Q_24hr_CNN_fused_smarp  
87   M_Q_24hr_CNN_fused_smarp  

[2 rows x 81 columns]

In [87]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

In [64]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_2[runs_2['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

## Compare leaderboard3: base vs base_rep_CNN
CNN on fused_sharp with seed 2 also proves AUC is a bad metric. High AUC may well associates with a low TSS in the under- or over-forecasting (if the threshold is 0.5)

In [90]:
runs_raw_1 = retrieve('leaderboard3', 'base')
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                          start_time tags.mlflow.runName  \
124 2021-08-18 03:00:51.425000+00:00                base   

                tags.mlflow.source.git.commit  
124  98d231a7d80767545d12d384c881f6eb545ffb7a  


In [91]:
runs_raw_2 = retrieve('leaderboard3', 'base_rep_CNN')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [92]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [93]:
runs_raw_3 = retrieve('leaderboard3', 'val_same_test')
runs_3 = select(runs_raw_3)
runs_3.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
83 2021-08-18 21:35:53.919000+00:00       val_same_test   

               tags.mlflow.source.git.commit  
83  717ab7690cf682989b9b190c2a8653f09ea4fdff  


In [94]:
runs_diff = diff(runs_1[runs_1['estimator'] == 'CNN'], runs_3[runs_3['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [95]:
runs_diff = diff(runs_2, runs_3[runs_3['estimator'] == 'CNN'])
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

## Compare leaderboard3: base_rep_CNN vs larger_lr
`Does larger learning rate help?`
* It doesn't even make sense to compare the performance if you don't select the robust model.
* It seems larger learning rate causes, though almost indistinguishable, slower convergence of the training loss (fused_sharp with seed 3 and 4, sharp with seed 3), which is interesting.

In [96]:
runs_raw_1 = retrieve('leaderboard3', 'base_rep_CNN')
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [97]:
runs_raw_2 = retrieve('leaderboard3', 'larger_lr')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
27 2021-08-19 08:03:56.682000+00:00           larger_lr   

               tags.mlflow.source.git.commit  
27  10eedfd9086bcc850e8538fa1ca9c6975495c544  


In [98]:
runs_diff = diff(runs_1[runs_1['dataset'].isin(['fused_sharp', 'sharp'])], runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

In [103]:
runs_comb = pd.concat((
    runs_raw_1.loc[[38]],
    runs_raw_2.loc[[15]],
))
runs_comb

run_id experiment_id    status  \
38  518b518563574a2b9a01a51bbe040ad9            15  FINISHED   
15  9d16191e1e84476dbeb2ded425b323ed            15  FINISHED   

                                                                                             artifact_uri  \
38  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/518b518563574a2b9a01a51bbe040ad9/artifacts   
15  file:///home/zeyusun/work/flare-prediction-smarp/mlruns/15/9d16191e1e84476dbeb2ded425b323ed/artifacts   

                         start_time                         end_time  \
38 2021-08-19 05:35:05.410000+00:00 2021-08-19 05:45:24.838000+00:00   
15 2021-08-19 11:00:01.676000+00:00 2021-08-19 12:48:34.576000+00:00   

    metrics.test/auc  metrics.test/recall  metrics.validation/accuracy  \
38          0.970089             0.789869                     0.908663   
15          0.970372             0.744841                     0.838041   

    metrics.test/hss1  ...  tags.mlflow.source.name  \
38           0.754221  ...           sweep_arnet.py   
15           0.729831  ...           sweep_arnet.py   

             tags.mlflow.parentRunId  tags.mlflow.source.type  tags.balanced  \
38  6f8716d1ce714738b3f524a87c37512b                    LOCAL           True   
15  e6062e9ba5354837846d0a656d39044f                    LOCAL           True   

    tags.mlflow.user  tags.database_name  \
38           zeyusun            M_Q_24hr   
15           zeyusun            M_Q_24hr   

               tags.mlflow.source.git.commit  tags.estimator_name  \
38  ed2b596b684b6bd9d96d47271a0e01b39da971d2                  CNN   
15  10eedfd9086bcc850e8538fa1ca9c6975495c544                  CNN   

                                                                                             tags.checkpoint  \
38   /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1350/checkpoints/epoch=7-step=439.ckpt   
15  /home/zeyusun/work/flare-prediction-smarp/lightning_logs/version_1371/checkpoints/epoch=15-step=879.ckpt   

    tags.mlflow.runName  
38   M_Q_24hr_CNN_sharp  
15   M_Q_24hr_CNN_sharp  

[2 rows x 81 columns]

In [105]:
dirs = tensorboard(runs_comb)
%tensorboard --logdir_spec {dirs}

## Compare leaderboard3: cnn_li2020

In [47]:
runs_raw_2 = retrieve('leaderboard3', 'cnn_li2020')
runs_2 = select(runs_raw_2)
runs_2.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
22 2021-08-19 09:04:34.552000+00:00          cnn_li2020   

               tags.mlflow.source.git.commit  
22  167b6d8cde9a134dba8cfe12a142cc97c97dcb58  


In [48]:
runs_raw_1 = retrieve('leaderboard3', 'base_rep_CNN')
runs_raw_1 = runs_raw_1[runs_raw_1['tags.dataset_name'].isin(['fused_sharp', 'sharp'])]
runs_1 = select(runs_raw_1)
runs_1.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
42 2021-08-19 05:07:03.892000+00:00        base_rep_CNN   

               tags.mlflow.source.git.commit  
42  ed2b596b684b6bd9d96d47271a0e01b39da971d2  


In [49]:
runs_diff = diff(runs_1, runs_2)
runs_diff.style.background_gradient(axis=0, vmin=-0.2, vmax=0.2, cmap='seismic')

* Code difference
|                  |   1: base_rep_CNN   |   2: cnn_li2020   |
|:----------------:|:-------------------:|:-----------------:|
|     patience     |          10         |         20        |
|   learning rate  |         1e-3        |        1e-2       |
| 1/5 of the other | used for validation | used for training |
* Performance difference
    * base_rep_CNN is very bad: fused_sharp's performance is all significantly worse than sharp's except for seed 1.
    * cnn_li2020: fused_sharp sees an increase in ACC but not AUC

## two_stages:test

In [56]:
runs_raw = retrieve('two_stages', 'test')
runs = select(runs_raw)
runs.style.background_gradient(axis=0)

Select iloc 0 from 
                         start_time tags.mlflow.runName  \
1  2021-08-22 06:36:45.309000+00:00                test   
3  2021-08-22 06:25:06.680000+00:00                test   
5  2021-08-22 06:22:07.482000+00:00                test   
8  2021-08-21 08:33:47.985000+00:00                test   
10 2021-08-21 08:28:55.500000+00:00                test   

               tags.mlflow.source.git.commit  
1   48c5794285044d9e7fe19bc0cb0238df3892da66  
3   48c5794285044d9e7fe19bc0cb0238df3892da66  
5   48c5794285044d9e7fe19bc0cb0238df3892da66  
8   1d5709bbc561def0dab992f1cd1f3d1a0acfdaa1  
10  1d5709bbc561def0dab992f1cd1f3d1a0acfdaa1  


In [57]:
dirs = tensorboard(runs_raw)
%tensorboard --logdir_spec {dirs}